### PR TITLE
adding a sign up screen and creating animations

### PR DESCRIPTION
**Pull request by shrevin**
created animations in login screen and sign up screen and created password constraints using regex

### CODE REVIEW

**Comment by shrevin** [comment]
ezgif com-gif-maker

File: Splash/Splash/Helper/Helper.m [hunk 1]
```diff
@@ -20,4 +21,47 @@ + (NSString*) formatTimeString:(int)secs {
...
+    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Missing Email" message:@"Please enter an email" preferredStyle:(UIAlertControllerStyleAlert)];
+        // create an OK action
+        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
+                                                           style:UIAlertActionStyleDefault
+                                                         handler:^(UIAlertAction * _Nonnull action) {
+                                                                 // handle response here.
+                                                         }];
+        // add the OK action to the alert controller
+        [alert addAction:okAction];
+        [[UIApplication sharedApplication].keyWindow.rootViewController presentViewController:alert animated:YES completion:^{}];
```
Comment (arthpatel29) [inline Splash/Splash/Helper/Helper.m]: This code is repeated 3 times, we can put all of them into one helper method and use it

File: Splash/Splash/Details/DetailsViewController.m [hunk 2]
```diff
@@ -24,6 +26,7 @@ @interface DetailsViewController () <UITableViewDelegate, UITableViewDataSource,
 @property (strong, nonatomic) PFUser *current;
 @property (strong, nonatomic) IBOutlet UIButton *onLBButton;
 @property (strong, nonatomic) NSArray *friends;
+@property ParseDataLoaderManager *dataLoader;
```
Comment (arthpatel29) [inline Splash/Splash/Details/DetailsViewController.m]: This should be id<DataLoaderProtocol> as discussed earlier.

File: Splash/Splash/Details/DetailsViewController.m [hunk 3]
```diff
@@ -50,10 +53,10 @@ - (void) setupTableView {
 }
 
 - (void) setupView {
-    self.usernameLabel.text = self.user[@"username"];
```
Comment (arthpatel29) [inline Splash/Splash/Details/DetailsViewController.m:53]: Divide this into smaller functions

File: Splash/Splash/Networking/ParseDataLoaderManager.m [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+- (void) loginUser: (NSString *)username password:(NSString *)password vc:(UIViewController *) vc segueId:(NSString *)segueId alert:(UIAlertController *)alert{
+    [PFUser logInWithUsernameInBackground:username password:password block:^(PFUser * user, NSError *  error) {
+        if (error != nil) {
+            DLog(@"User log in failed: %@", error.localizedDescription);
+            [[UIApplication sharedApplication].keyWindow.rootViewController presentViewController:alert animated:YES completion:^{}];
+        } else {
+            DLog(@"User logged in successfully");
+            [vc performSegueWithIdentifier:@"segueFromLogin" sender:self];
+            // display view controller that needs to shown after successful login
+        }
```
Comment (arthpatel29) [inline Splash/Splash/Networking/ParseDataLoaderManager.m]: Put this inside completion handler

File: Splash/Splash/Networking/DataLoaderProtocol.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+- (void) updateGoal: (NSDate *)goal;
+- (NSMutableArray *) getLeaderboardData;
+- (NSDate *) getGoalAsDate: (PFUser *) user;
+- (void) updateBubblescore: (PFUser *) user newScore:(int)newScore;
+- (void) updateStreak: (PFUser *) user newStreak:(int)newStreak;
+- (void) updateTotalShowerTime: (PFUser *) user newTime:(int)newTime;
+- (void) updateNumShowers: (PFUser *) user newNum:(int)newNum;
+- (NSArray *) getFriends: (PFUser *) user;
+- (NSArray *) getPossibleNewFriends: (NSArray*)currFriends;
+- (void) getShowerData: (NSMutableArray *) originalArray filteredArray:(NSMutableArray *)filteredArray tableView:(UITableView *)tableView;
```
Comment (arthpatel29) [inline Splash/Splash/Networking/DataLoaderProtocol.h]: Lets group together similar functions

File: Splash/Splash/Networking/ParseDataLoaderManager.m [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    [imageView loadInBackground];
+}
+
+// getting the current user's goal stat
+- (NSString *) getGoal: (PFUser *)user {
+    NSCalendar *calendar = [NSCalendar currentCalendar];
+    NSDateComponents *components = [calendar components:(NSCalendarUnitMinute | NSCalendarUnitSecond) fromDate:[PFUser currentUser][@"goal"]];
+    int currMinute = [components minute];
+    int currSeconds = [components second];
+    return [NSString stringWithFormat:@"%@", [Helper formatTimeString:(currMinute*60) + currSeconds]];
```
Comment (arthpatel29) [inline Splash/Splash/Networking/ParseDataLoaderManager.m]: Lets refactor this.

File: Splash/Splash/Login/LoginViewController.m [hunk 7]
```diff
@@ -85,27 +77,38 @@ - (void) changeImage {
...
-                                                         }];
+    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
+        // handle response here.
+    }];
     // add the OK action to the alert controller
     [alert addAction:okAction];
     NSString *username = self.usernameField.text;
     NSString *password = self.passwordField.text;
-    [self.dataLoader loginUser:username password:password vc:self segueId:@"segueFromLogin" alert:alert];
+    [self.dataLoader loginUser:username password:password completion:^(PFUser *user, NSError *error) {
```
Comment (arthpatel29) [inline Splash/Splash/Login/LoginViewController.m]: Is user parameter needed in the completion

File: Splash/Splash/Routine/RoutineViewController.swift [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        textfield3.textAlignment = NSTextAlignment.center
+        subview.addSubview(textfield3)
+                
+        // Add the subview to the alert's UI property
+        alert.customSubview = subview
+        _ = alert.addButton("Add") {
+            print("Clicked Add")
+            print(textfield1.text)
+            print(textfield2.text)
+            print(textfield3.text)
```
Comment (arthpatel29) [inline Splash/Splash/Routine/RoutineViewController.swift]: let's remove print

File: Splash/Splash/Routine/RoutineViewController.swift [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        subview.addSubview(textfield2)
+        
+        // Add textfield 3
+        let textfield3 = UITextField(frame: CGRect(x: x,y: textfield2.frame.maxY + 10,width: 180,height: 35))
+        textfield3.layer.borderWidth = 1.5
+        textfield3.layer.cornerRadius = 5
+        textfield3.layer.borderColor = UIColor.black.cgColor
+        textfield3.placeholder = "# of seconds"
+        textfield3.textAlignment = NSTextAlignment.center
+        subview.addSubview(textfield3)
```
Comment (arthpatel29) [inline Splash/Splash/Routine/RoutineViewController.swift]: We can move this repeated code to a helper method

File: Splash/Splash/Details/DetailsViewController.m [hunk 10]
```diff
@@ -24,6 +26,7 @@ @interface DetailsViewController () <UITableViewDelegate, UITableViewDataSource,
 @property (strong, nonatomic) PFUser *current;
 @property (strong, nonatomic) IBOutlet UIButton *onLBButton;
 @property (strong, nonatomic) NSArray *friends;
+@property id <DataLoaderProtocol> dataLoader;
```
Comment (yaxuanwang) [inline Splash/Splash/Details/DetailsViewController.m:29]: Shall we add attributes for this property?

File: Splash/Splash/Friends/FriendsCell.h [hunk 11]
```diff
@@ -8,13 +8,16 @@
...
+#import "DataLoaderProtocol.h"
+#import "ParseDataLoaderManager.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
 @interface FriendsCell : UICollectionViewCell
 @property (strong, nonatomic) IBOutlet UILabel *friendName;
 @property (strong, nonatomic) IBOutlet PFImageView *friendPic;
 @property (strong, nonatomic) PFUser *user;
+@property ParseDataLoaderManager *dataLoader;
```
Comment (yaxuanwang) [inline Splash/Splash/Friends/FriendsCell.h:20]: same here

File: Splash/Splash/Friends/FriendsViewController.m [hunk 12]
```diff
@@ -17,19 +19,28 @@ @interface FriendsViewController () <UICollectionViewDelegate, UICollectionViewD
...
 @property (strong, nonatomic) IBOutlet UISearchBar *searchBar;
+@property ParseDataLoaderManager *dataLoader;
 @end
 
 @implementation FriendsViewController
 
 - (void)viewDidLoad {
     [super viewDidLoad];
     // Do any additional setup after loading the view.
+    self.dataLoader = [[ParseDataLoaderManager alloc] init];
```
Comment (yaxuanwang) [inline Splash/Splash/Friends/FriendsViewController.m:30]: nit: just use [ParseDataLoaderManager new]?

File: Splash/Splash/Impact/ImpactViewController.m [hunk 13]
```diff
@@ -54,15 +54,13 @@ - (void) dismissKeyboard
...
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
     // Get the new view controller using [segue destinationViewController].
-    // Pass the selected object to the new view controller.
-//    ImpactCalculationsViewController *impactCalc = [segue destinationViewController];
-//    impactCalc.waterFlow = [self.waterFlowTextField.text floatValue];
-//    impactCalc.showersPerWeek = [self.showersTextField.text intValue];
 }
```
Comment (yaxuanwang) [inline Splash/Splash/Impact/ImpactViewController.m:63]: Do you still need this function?

File: Splash/Splash/Routine/RoutineCell.m [hunk 14]
```diff
@@ -0,0 +1,12 @@
+//
+//  RoutineCell.m
+//  Splash
+//
+//  Created by Shreya Vinjamuri on 8/2/22.
+//
+
+#import "RoutineCell.h"
+
+@implementation RoutineCell
```
Comment (yaxuanwang) [inline Splash/Splash/Routine/RoutineCell.m:10]: Do you plan to add more implementation here as follow up?

Comment (shrevin) [reply Splash/Splash/Routine/RoutineCell.m:10]: Yes, I have.

**Comment by shrevin** [comment]
routine feature

**Pull request merged**